src/frontend/circom/mod.rs

```diff
@@ -0,0 +1,229 @@
+use std::{error::Error, fs::File, io::BufReader, path::Path};
+
+use color_eyre::Result;
+use num_bigint::BigInt;
+
+use ark_circom::{circom::r1cs_reader, WitnessCalculator};
+use ark_ec::pairing::Pairing;
+use ark_ff::PrimeField;
+
+use crate::ccs::r1cs::R1CS;
+use crate::utils::vec::SparseMatrix;
+
+// Define the sparse matrices on PrimeFiled
+pub type Constraints<F> = (ConstraintVec<F>, ConstraintVec<F>, ConstraintVec<F>);
+pub type ConstraintVec<F> = Vec<(usize, F)>;
+
+type ExtractedConstraintsResult<F> = Result<(Vec<Constraints<F>>, usize), Box<dyn Error>>;
+
+// Convert the generic type of constraints from Pairing's ScalarField to PrimeField.
+pub fn convert_constraints_bigint_to_scalar<E, F>(
+    constraints: Constraints<E::ScalarField>,
+) -> Constraints<F>
+where
+    E: Pairing,
+    F: PrimeField,
+{
+    let convert_vec = |vec: ConstraintVec<E::ScalarField>| -> ConstraintVec<F> {
+        vec.into_iter()
+            .map(|(index, element)| {
+                // Convert element into BigInt, then to BigUint, then to the destination PrimeField
+                let bigint_e: <<E as Pairing>::ScalarField as PrimeField>::BigInt =
+                    element.into_bigint();
+                let generic_biguint: num_bigint::BigUint = bigint_e.into();
+                let f_element: F = F::from(generic_biguint);
```

Comment:
Do you think that we can get rid of the `F` since would be the same as `E::ScalarField`?

Reply:
I also feel this function seems unnecessary, I think we could have `extract_constraints_from_r1cs` to be like:

```rust
pub fn extract_constraints_from_r1cs<E>(
    r1cs_filepath: &Path,
) -> ExtractedConstraintsResult<E::ScalarField>
where
    E: Pairing,
{
    // ...
}
```

Reply:
I don't like this either.
We have one function which body is one closure.
Couldn't we do it on another way?

I think @han0110 's suggestion looks good though

Reply:
Thank you for the feedback. I've removed the convert_constraints_bigint_to_scalar function as it indeed was completely unnecessary. The code is now simplified by using only E::ScalarField.